### PR TITLE
Add support for NXP KW41Z4

### PR DESCRIPTION
**Pull request by georgecpr**

### CODE REVIEW

File: examples/platforms/kw41z/platform.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    temp = RSIM->CONTROL;
+    temp &= ~RSIM_CONTROL_RF_OSC_EN_MASK;
+    RSIM->CONTROL = temp | RSIM_CONTROL_RF_OSC_EN(1);
+    /* Prevent XTAL_OUT_EN from generating XTAL_OUT request */
+    RSIM->RF_OSC_CTRL |= RSIM_RF_OSC_CTRL_RADIO_EXT_OSC_OVRD_EN_MASK;
+
+    /* wait for RF_OSC_READY */
+    while ((RSIM->CONTROL & RSIM_CONTROL_RF_OSC_READY_MASK) == 0) {}
+
+    if (0 == revId)
```
Comment (aeliot) [inline examples/platforms/kw41z/platform.c]: The variable should be on the left hand side of the conditional.
```
if (revId == 0)
```

File: examples/platforms/kw41z/flash.c [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+ThreadError utilsFlashErasePage(uint32_t aAddress)
+{
+    ThreadError error;
+    status_t status;
+
+    status = FLASH_Erase(&sFlashConfig, aAddress, FSL_FEATURE_FLASH_PFLASH_BLOCK_SECTOR_SIZE, kFLASH_ApiEraseKey);
+
+    if (kStatus_FLASH_Success == status)
```
Comment (aeliot) [inline examples/platforms/kw41z/flash.c]: Variable should be on left side of statement
```
if (status == kStatus_FLASH_Success)
```

File: examples/platforms/kw41z/platform.c (hunk 1 above)
Comment (georgecpr) [inline examples/platforms/kw41z/platform.c]: Fixed.

File: examples/platforms/kw41z/flash.c (hunk 2 above)
Comment (georgecpr) [inline examples/platforms/kw41z/flash.c]: Fixed.

**Comment by beriberikix** [comment]
@googlebot @georgecpr is covered by Corporate CLA signed by NXP Semiconductors.

File: examples/platforms/kw41z/openthread-core-kw41z-config.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+/**
+ * @file
+ *   This file includes kw41z compile-time configuration constants
+ *   for OpenThread.
+ */
+
+#ifndef OPENTHREAD_CORE_KW41Z_CONFIG_H_
+#define OPENTHREAD_CORE_KW41Z_CONFIG_H_
```
Comment (nibanks) [inline examples/platforms/kw41z/openthread-core-kw41z-config.h]: From `radio.c` it looks like the only hw offload you support is energy scan (similar to NRF52840), so you should define the following here to allow for the raw link layer to expose full functionality:

```
 /**
  * @def OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
  *
  * Define to 1 if you want to enable software ACK timeout logic.
  *
  */
#define OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT           1

 /**
  * @def OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT
  *
  * Define to 1 if you want to enable software retransmission logic.
  *
  */
#define OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT            1
```

Comment (georgecpr) [inline examples/platforms/kw41z/openthread-core-kw41z-config.h]: The re-transmissions should be handled by SW, but ACK time-out is handled by the radio.
I will enable the OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT config option.

File: examples/platforms/kw41z/radio.c [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+static ThreadError  rf_remove_addr_table_entry(uint16_t checksum);
+static ThreadError  rf_remove_addr_table_entry_index(uint8_t index);
+
+PhyState otPlatRadioGetState(otInstance *aInstance)
+{
+    (void)aInstance;
+    return sState;
+}
+
+void otPlatRadioGetIeeeEui64(otInstance *aInstance, uint8_t *aIeeeEui64)
```
Comment (xiaom-GitHub) [inline examples/platforms/kw41z/radio.c]: When I try with one kw41z board, seems this function will cause reset the board. Could you have a check here?

File: examples/platforms/kw41z/radio.c [hunk 5]
```diff
@@ -0,0 +1,900 @@
...
+        pSrc = (uint8_t *)&SIM->UIDML;
+    }
+    else
+    {
+        pSrc = (uint8_t *)&RSIM->MAC_MSB;
+    }
+
+    for (i = 0; i < 8; i++)
+    {
+        aIeeeEui64[i] = pSrc[7 - i];
```
Comment (xiaom-GitHub) [inline examples/platforms/kw41z/radio.c]: nit: magic number?

File: examples/platforms/kw41z/radio.c (hunk 4 above)
Comment (georgecpr) [inline examples/platforms/kw41z/radio.c]: Fixed.

**Comment by chshu** [comment]
@georgecpr, Would you mind to take a look at #1786.

I noticed that current KW41Z4 radio driver doesn't not pass up the ACK packet, but only pass up the frame pending flag. To support #1786, Is it possible to update current implementation and pass up the whole ACK packet? Would be Similar to some other OT platforms, like cc2538/efr32/da15000.

**Comment by georgecpr** [comment]
@chshu The current KW41Z driver does not passes the ACK frame to the upper layer because the hardware is handling these frames (checks the received frame type, seqNo, CRC) and does not store them in the  packet buffer.
It is possible to pass these frames to the upper layer, but some HW acceleration needs to be disabled, and additional SW checks needs to be added.

**Comment by chshu** [comment]
@georgecpr, Thanks for the information. I think for short-term, it could be a interim solution to support two different `TransmitDone()` callbacks (so platforms could determine passing up ACK or not by using different callback), but eventually all platforms need to pass up ACK frame to support Enhanced ACK feature.

File: examples/platforms/kw41z/MKW41Z512xxx4.ld [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  /* Initializes stack on the end of block */
+  __StackTop   = ORIGIN(m_data) + LENGTH(m_data);
+  __StackLimit = __StackTop - STACK_SIZE;
+  PROVIDE(__stack = __StackTop);
+
+  .ARM.attributes 0 : { *(.ARM.attributes) }
+
+  ASSERT(__StackLimit >= __HeapLimit, "region m_data overflowed with stack and heap")
+}
+
```
Comment (xiaom-GitHub) [inline examples/platforms/kw41z/MKW41Z512xxx4.ld]: nit: extra line?

File: examples/platforms/kw41z/Makefile.am [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    @top_builddir@/third_party/nxp/MKW41Z4/XCVR/cfgs_kw4x_3x_2x/fsl_xcvr_zgbe_config.c               \
+    $(NULL)
+
+if OPENTHREAD_ENABLE_DIAG
+PLATFORM_SOURCES                                                                                  += \
+    diag.c                                                                                           \
+    $(NULL)
+endif
+
+
```
Comment (xiaom-GitHub) [inline examples/platforms/kw41z/Makefile.am]: nit: extra line?

File: examples/platforms/kw41z/README.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+ ```bash
+ > ping fdde:ad00:beef:0:0:ff:fe00:fc00
+ 16 bytes from fdde:ad00:beef:0:0:ff:fe00:fc00: icmp_seq=1 hlim=64 time=8ms
+ ```
+
+For a list of all available commands, visit [OpenThread CLI Reference README.md][CLI].
+
+[CLI]: https://github.com/openthread/openthread/blob/master/src/cli/README.md
+
```
Comment (xiaom-GitHub) [inline examples/platforms/kw41z/README.md]: nit: extra line?

File: examples/platforms/kw41z/MKW41Z512xxx4.ld (hunk 6 above)
Comment (georgecpr) [inline examples/platforms/kw41z/MKW41Z512xxx4.ld]: fixed

File: examples/platforms/kw41z/Makefile.am (hunk 7 above)
Comment (georgecpr) [inline examples/platforms/kw41z/Makefile.am]: fixed

File: examples/platforms/kw41z/README.md (hunk 8 above)
Comment (georgecpr) [inline examples/platforms/kw41z/README.md]: fixed

**Comment by beriberikix** [comment]
@georgecpr if desired, please add NXP to the AUTHORS file.

**Pull request merged**